### PR TITLE
Client version 6.1.7

### PR DESCRIPTION
**Pull request by idantavor**

### CODE REVIEW

File: test/aerospike_clj/integration_test.clj [hunk 1]
```diff
@@ -529,15 +529,14 @@
...
-    (is (= ReadModeAP/ONE (.readModeAP rp))) ;; Involve single node in the read operation.
+     ;; Involve single node in the read operation.
     (is (= Replica/SEQUENCE (.replica rp))) ;; Try node containing master partition first.
     ;; If connection fails, all commands try nodes containing replicated partitions.
     ;; If socketTimeout is reached, reads also try nodes containing replicated partitions,
     ;; but writes remain on master node.)))
     ;; This option requires ClientPolicy.requestProleReplicas to be enabled in order to function properly.
     (is (= 30000 (.socketTimeout rp))) ;; 30 seconds default
-    (is (zero? (.totalTimeout rp))) ;; no time limit
+    (is (= 1000 (.totalTimeout rp))) ;; no time limit
```
Comment (evg-tso) [inline test/aerospike_clj/integration_test.clj]: Is the comment still accurate?

File: test/aerospike_clj/integration_test.clj [hunk 2]
```diff
@@ -586,8 +584,9 @@
     ;; If socketTimeout is reached, reads also try nodes containing replicated partitions,
     ;; but writes remain on master node.)))
     ;; This option requires ClientPolicy.requestProleReplicas to be enabled in order to function properly.
+    ;; This option requires ClientPolicy.requestProleReplicas to be enabled in order to function properly.
     (is (= 30000 (.socketTimeout rp))) ;; 30 seconds default
-    (is (zero? (.totalTimeout rp))) ;; no time limit
+    (is (= 1000 (.totalTimeout rp))) ;; no time limit
```
Comment (evg-tso) [inline test/aerospike_clj/integration_test.clj]: Is the comment still accurate?

File: test/aerospike_clj/integration_test.clj [hunk 3]
```diff
@@ -529,15 +529,14 @@
 
 (deftest default-read-policy
   (let [rp (.getReadPolicyDefault ^AerospikeClient (.-client ^SimpleAerospikeClient *c*))]
-    (is (= Priority/DEFAULT (.priority rp))) ;; Priority of request relative to other transactions. Currently, only used for scans.
-    (is (= ReadModeAP/ONE (.readModeAP rp))) ;; Involve single node in the read operation.
+     ;; Involve single node in the read operation.
```
Comment (evg-tso) [inline test/aerospike_clj/integration_test.clj]: Should this comment be deleted?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+on: [workflow_call]
+
+jobs:
+  test:
+    runs-on: ubuntu-22.04
+    timeout-minutes: 5
+    services:
+      aerospike-test:
+        image: aerospike:ee-6.2.0.3
```
Comment (evg-tso) [inline .github/workflows/test.yml]: Should we update the README what's the minimum supported version?

File: test/aerospike_clj/integration_test.clj [hunk 5]
```diff
@@ -14,7 +14,7 @@
   (:import [com.aerospike.client Value AerospikeClient]
            [com.aerospike.client.cdt ListOperation ListPolicy ListOrder ListWriteFlags ListReturnType
                                      MapOperation MapPolicy MapOrder MapWriteFlags MapReturnType CTX]
-           [com.aerospike.client.policy Priority ReadModeSC ReadModeAP Replica GenerationPolicy RecordExistsAction
+§           [com.aerospike.client.policy ReadModeSC ReadModeAP Replica GenerationPolicy RecordExistsAction
```
Comment (evg-tso) [inline test/aerospike_clj/integration_test.clj]:
```suggestion
           [com.aerospike.client.policy ReadModeSC ReadModeAP Replica GenerationPolicy RecordExistsAction
```

File: .github/workflows/test.yml (hunk 4 above)
Comment (evg-tso) [inline .github/workflows/test.yml]: Please also update [this docstring](https://github.com/AppsFlyer/aerospike-clj/pull/62/files#diff-74af9967e37bd46c41423e45248ecec8f56eced19e3ac284951423d19d955455R6) to match the new Aerospike version.

File: README.md [hunk 6]
```diff
@@ -16,6 +16,7 @@ An opinionated Clojure library wrapping Aerospike Java Client.
 # Requirements
 - Java 8
 - Clojure 1.8
+- Aerospike server version >= `4.9.0`
```
Comment (evg-tso) [inline README.md:19]: Does this mean that a client version `6.1.10` can work with a server version of `4.9.0`?

File: README.md [hunk 7]
```diff
@@ -55,17 +56,19 @@ user=> (def c (aero/init-simple-aerospike-client
 It is possible to inject additional asynchronous user-defined behaviour. To do that add an implementation of  the
 `ClientEvents` protocol.
 Some useful info is passed in-order to support metering and to read client configuration. `op-start-time` is
-`(System/nanoTime)`, see more [here](https://appsflyer.github.io/aerospike-clj/advanced-async-hooks.html).
+`(System/nanoTime)`.
+Additional context can be supplied by attaching a map to the config object under `:client-events-ctx` key. 
+see more [here](https://appsflyer.github.io/aerospike-clj/advanced-async-hooks.html).
```
Comment (evg-tso) [inline README.md:61]: Perhaps we should update that link, it still refers to `manifold`.

File: README.md [hunk 8]
```diff
@@ -55,17 +56,19 @@ user=> (def c (aero/init-simple-aerospike-client
...
           ["localhost"]
           "test"
           {:client-events (reify ClientEvents
-                            (on-success [_ op-name op-result index op-start-time db]
-                              (when (:enable-logging? db)
+                            (on-success [_ op-name op-result index op-start-time ctx]
+                              (when (-> ctx :enable-logging?)
                                 (println op-name "success!")))
-                            (on-failure [_  op-name op-ex index op-start-time db]
+                            (on-failure [_  oop-name op-result index op-start-time ctx]
```
Comment (evg-tso) [inline README.md]: 1. Should `oop-name` be `op-name`?
2. Why did you rename `op-ex` to `op-result`?

File: src/main/clojure/aerospike_clj/client.clj [hunk 9]
```diff
@@ -12,17 +13,19 @@
...
                                         RecordExistsAction WritePolicy ScanPolicy
                                         InfoPolicy]
-           [com.aerospike.client Key Host]
+           [com.aerospike.client Key Host BatchRecord]
            [aerospike_clj.listeners AsyncExistsListener AsyncDeleteListener AsyncWriteListener
                                     AsyncInfoListener AsyncRecordListener AsyncRecordSequenceListener
-                                    AsyncBatchListListener AsyncExistsArrayListener]))
+                                    AsyncBatchListListener AsyncExistsArrayListener AsyncBatchOperateListListener]
+           (com.aerospike.client.listener BatchOperateListListener)
+           (java.util.concurrent Executor)))
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj:28]: This is me being pedantic, but the newer imports use `(` and `)` (like they should), but the older ones use `[` and `]`.  
Would you mind aligning them?

File: src/main/clojure/aerospike_clj/client.clj [hunk 10]
```diff
@@ -54,18 +57,18 @@
   (let [elp (policy/map->event-policy conf)]
     (NioEventLoops. elp 1 true "NioEventLoops")))
 
-(defn- client-events-reducer [op-name index op-start-time]
+(defn- client-events-reducer [op-name index op-start-time ctx]
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Perhaps we should we consistent with either `ctx` or `client-events-ctx`.

File: src/main/clojure/aerospike_clj/client.clj [hunk 11]
```diff
@@ -133,10 +138,10 @@
               ^Policy (:policy conf)
               ^Key (pt/create-key index dbns set-name)
               ^"[Ljava.lang.String;" (utils/v->array String bin-names)))
-      (let [p (p/chain op-future
-                       record/record->map
-                       (:transcoder conf identity))]
-        (register-events p client-events :read index start-time))))
+      (-> op-future
+          (p/then' record/record->map completion-executor)
+          (p/then' (:transcoder conf identity))
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Can we combine these two `then'` into a single one?

File: src/main/clojure/aerospike_clj/client.clj [hunk 12]
```diff
@@ -155,7 +160,9 @@
                (AsyncExistsListener. op-future)
                ^Policy (:policy conf)
                ^Key (pt/create-key index dbns set-name))
-      (register-events op-future client-events :exists index start-time)))
+      (-> op-future
+          (p/then identity completion-executor)
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: I think `then'` will work better here.

File: src/main/clojure/aerospike_clj/protocols.clj [hunk 13]
```diff
@@ -154,9 +164,9 @@
 (defprotocol ClientEvents
   "Continuation functions that are registered when an async DB operation is called.
   The value returned from those function will be the value of the returned future from the async operation."
-  (on-success [this op-name op-result index op-start-time]
+  (on-success [this op-name op-result index op-start-time ctx]
```
Comment (asafch) [inline src/main/clojure/aerospike_clj/protocols.clj]: I think that if we're going in the direction of a `ctx` then we might as well wrap everything in a map and pass it, so the protocol would be:
```clojure
(defprotocol ClientEvents
  (on-success [this ctx])
  (on-failure [this ctx]))
```

Where we guarantee a minimal schema of `ctx` in each case:

- For `on-success` we guarantee `op-name`, `op-result`, `index`, `op-start-time`
- For `on-failure` we guarantee `op-name`, `op-ex`, `index`, `op-start-time`

But, this would cause more short-lived objects and a lot more GC.

File: README.md (hunk 6 above)
Comment (idantavor) [reply README.md:19]: yes

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,6 +1,19 @@
 ## This library follows [Semantic Versioning](https://semver.org).
 ## This CHANGELOG follows [keepachangelog](https://keepachangelog.com/en/1.0.0/).
 
+###  VERSION 3.0.0
```
Comment (evg-tso) [inline CHANGELOG.md]: Would you mind adding a link to the PR?

Add this at the end of the file:
```
[3.0.0]: https://github.com/AppsFlyer/aerospike-clj/pull/62
```

And a link here:
```suggestion
###  VERSION [3.0.0]
```

File: src/main/clojure/aerospike_clj/client.clj (hunk 9 above)
Comment (evg-tso) [reply src/main/clojure/aerospike_clj/client.clj:28]: Like this:
```clojure
(ns aerospike-clj.client
  (:refer-clojure :exclude [update])
  (:require [aerospike-clj.aerospike-record :as record]
            [aerospike-clj.bins :as bins]
            [aerospike-clj.key :as as-key]
            [aerospike-clj.listeners]
            [aerospike-clj.metrics :as metrics]
            [aerospike-clj.policy :as policy]
            [aerospike-clj.protocols :as pt]
            [aerospike-clj.utils :as utils]
            [clojure.string :as s]
            [clojure.tools.logging :as log]
            [promesa.core :as p]
            [promesa.exec :as p-exec])
  (:import (aerospike_clj.listeners AsyncBatchListListener AsyncBatchOperateListListener AsyncDeleteListener
                                    AsyncExistsArrayListener AsyncExistsListener AsyncInfoListener
                                    AsyncRecordListener AsyncRecordSequenceListener AsyncWriteListener)
           (com.aerospike.client BatchRecord Host Key)
           (com.aerospike.client AerospikeClient BatchRead Bin Key Operation)
           (com.aerospike.client.async EventLoop EventLoops NioEventLoops)
           (com.aerospike.client.cluster Node)
           (com.aerospike.client.listener BatchOperateListListener)
           (com.aerospike.client.policy BatchPolicy ClientPolicy InfoPolicy
                                        Policy RecordExistsAction ScanPolicy
                                        WritePolicy)
           (java.time Instant)
           (java.util ArrayList Arrays Collection List)
           (java.util.concurrent Executor)))
```

File: test/aerospike_clj/integration/aerospike_setup.clj [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+                       :message         "objects: all 0 master 0 prole 0 non-replica 0"
+                       :times           1
+                       :startup-timeout 15}}
+      tc/create
+      tc/start!
+      (reset! container)))
+
+(defn- stop! []
+  (tc/stop! @container)
+  (reset! container nil?))
```
Comment (evg-tso) [inline test/aerospike_clj/integration/aerospike_setup.clj]:
```suggestion
  (reset! container nil))
```

File: src/main/clojure/aerospike_clj/aerospike_record.clj [hunk 16]
```diff
@@ -14,15 +14,16 @@
...
-             payload   (if (single-bin? bins)
-                         ;; single bin record
-                         (utils/desanitize-bin-value (.get bins ""))
-                         ;; multiple-bin record
-                         (reduce-kv (fn [m k v]
-                                      (assoc m k (utils/desanitize-bin-value v)))
-                                    {}
-                                    bins))]
+       (let [bins    ^Map (.bins record)
+             payload (and bins
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/aerospike_record.clj]: I know that `and` "returns the value of the last expr. (and) returns true".  
But, it's less readable IMO.

File: src/main/clojure/aerospike_clj/client.clj [hunk 17]
```diff
@@ -133,10 +139,13 @@
...
-      (let [p (p/chain op-future
-                       record/record->map
-                       (:transcoder conf identity))]
-        (register-events p client-events :read index start-time))))
+      (-> op-future
+          (p/then' (fn [res]
+                     (let [transcoder (:transcoder conf identity)]
+                       (-> res
+                           record/record->map
+                           transcoder))) completion-executor)
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Please put the `completion-executor` in a separate line to increase readability.

File: src/main/clojure/aerospike_clj/client.clj [hunk 18]
```diff
@@ -186,10 +197,10 @@
                (AsyncExistsArrayListener. op-future)
                ^BatchPolicy (:policy conf)
                ^"[Lcom.aerospike.client.Key;" indices)
-      (let [d (p/chain op-future
-                       vec
-                       (:transcoder conf identity))]
-        (register-events d client-events :exists-batch nil start-time))))
+      (-> op-future
+          (p/then' vec completion-executor)
+          (p/then' (:transcoder conf identity))
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Can the `transcoder` be called in the previous `p/then'`, removing one continuation callback?

Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Same question about the `transcoder` and using a single `p/then'`

File: src/main/clojure/aerospike_clj/client.clj [hunk 19]
```diff
@@ -288,7 +305,9 @@
               (AsyncWriteListener. op-future)
               ^WritePolicy (policy/write-policy client expiration RecordExistsAction/UPDATE_ONLY)
               ^Key (pt/create-key index dbns set-name))
-      (register-events op-future client-events :touch index start-time)))
+      (-> op-future
+          (p/then' identity completion-executor)
+          (register-events client-events :touch index start-time {}))))
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: We don't want to give the clients a `conf` parameter?

File: src/main/clojure/aerospike_clj/client.clj [hunk 20]
```diff
@@ -335,7 +358,33 @@
...
+                            (Arrays/asList)))
+          start-time (System/nanoTime)]
+      (.operate ^AerospikeClient client
+                ^EventLoop (.next ^EventLoops el)
+                ^BatchOperateListListener (AsyncBatchOperateListListener. op-future)
+                ^BatchPolicy policy
+                ^List batch-list)
+      (-> op-future
+          (p/then' #(mapv batch-record->map %) completion-executor)
+          (p/then' (:transcoder conf identity))
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/client.clj]: Same question about the `:transcoder`

File: src/main/clojure/aerospike_clj/listeners.clj [hunk 21]
```diff
@@ -4,7 +4,7 @@
   (:import [java.util List Map]
```
Comment (evg-tso) [inline src/main/clojure/aerospike_clj/listeners.clj]: Would you mind aligning the `:import` to use `()` instead of `[]`?

File: test/aerospike_clj/integration/aerospike_setup.clj [hunk 22]
```diff
@@ -0,0 +1,31 @@
+(ns aerospike-clj.integration.aerospike-setup
+  (:require [clj-test-containers.core :as tc]))
+
+(def ^:private ^:const container-port 3000)
+(def ^:private container (atom nil))
```
Comment (evg-tso) [inline test/aerospike_clj/integration/aerospike_setup.clj]: Since this is a private `def`, I think you can use a dynamic var instead, and use `binding` instead of `reset!`

**Review by evg-tso — APPROVED** [review]

File: src/main/clojure/aerospike_clj/client.clj (hunk 20 above)
Comment (idantavor) [reply src/main/clojure/aerospike_clj/client.clj]: do you think it's better to compose it with the mapv?
I think we should think about moving the transcoder all together, but then you can do it in the same completion stage.

File: README.md (hunk 7 above)
Comment (idantavor) [reply README.md:61]: The actual doc is updated in the repo but the update-script fails when trying to use codox.
